File: qadevOOo/tests/java/mod/_sc/ScDataPilotFieldGroupItemObj.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */
package mod._sc;

import com.sun.star.beans.XPropertySet;
import com.sun.star.container.XIndexAccess;
import com.sun.star.container.XNameAccess;
import com.sun.star.container.XNamed;
import com.sun.star.lang.XComponent;
import com.sun.star.sheet.DataPilotFieldGroupInfo;
import com.sun.star.sheet.XDataPilotDescriptor;
import com.sun.star.sheet.XDataPilotFieldGrouping;
import com.sun.star.sheet.XDataPilotTables;
import com.sun.star.sheet.XDataPilotTablesSupplier;
import com.sun.star.sheet.XSpreadsheet;
import com.sun.star.sheet.XSpreadsheetDocument;
import com.sun.star.sheet.XSpreadsheets;
import com.sun.star.table.CellAddress;
import com.sun.star.table.CellRangeAddress;
import com.sun.star.uno.AnyConverter;
import com.sun.star.uno.Type;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XInterface;

import java.io.PrintWriter;

import lib.TestCase;
import lib.TestEnvironment;
import lib.TestParameters;

import util.SOfficeFactory;



public class ScDataPilotFieldGroupItemObj extends TestCase
{
    private XSpreadsheetDocument xSheetDoc = null;

    /**
     * A field is filled some values. This integer determines the size of the
     * field in x and y direction.
     */
    private static final int mMaxFieldIndex = 6;

    /**
     * Creates Spreadsheet document.
     */
    @Override
    protected void initialize (TestParameters tParam, PrintWriter log) throws Exception
    {
        SOfficeFactory SOF = SOfficeFactory.getFactory (
            tParam.getMSF ());

        log.println ("creating a Spreadsheet document");
        xSheetDoc = SOF.createCalcDoc (null);
    }

    /**
     * Disposes Spreadsheet document.
     */
    @Override
    protected void cleanup (TestParameters tParam, PrintWriter log)
    {
        log.println ("    disposing xSheetDoc ");

        XComponent oComp = UnoRuntime.queryInterface (
            XComponent.class, xSheetDoc);
        util.DesktopTools.closeDoc (oComp);
    }

    @Override
    protected TestEnvironment createTestEnvironment (TestParameters Param,
        PrintWriter log) throws Exception
    {
        XInterface oObj = null;
        XInterface datapilotfield = null;
        XInterface groups = null;

        // creation of testobject here
        // first we write what we are intend to do to log file
        log.println ("Creating a test environment");

        // the cell range
        CellRangeAddress sCellRangeAdress = new CellRangeAddress ();
        sCellRangeAdress.Sheet = 0;
        sCellRangeAdress.StartColumn = 1;
        sCellRangeAdress.StartRow = 0;
        sCellRangeAdress.EndColumn = mMaxFieldIndex - 1;
        sCellRangeAdress.EndRow = mMaxFieldIndex - 1;

        // position of the data pilot table
        CellAddress sCellAdress = new CellAddress ();
        sCellAdress.Sheet = 0;
        sCellAdress.Column = 7;
        sCellAdress.Row = 8;

        log.println ("Getting a sheet");

        XSpreadsheets xSpreadsheets = xSheetDoc.getSheets ();
        XSpreadsheet oSheet = null;
        XSpreadsheet oSheet2 = null;
        XIndexAccess oIndexAccess = UnoRuntime.queryInterface (
            XIndexAccess.class, xSpreadsheets);
        // Make sure there are at least two sheets
        xSpreadsheets.insertNewByName("Some Sheet", (short)0);

        oSheet = (XSpreadsheet) AnyConverter.toObject (
            new Type (XSpreadsheet.class),
            oIndexAccess.getByIndex (0));
        oSheet2 = (XSpreadsheet) AnyConverter.toObject (
            new Type (XSpreadsheet.class),
            oIndexAccess.getByIndex (1));

        log.println ("Filling a table");

        for (int i = 1; i < mMaxFieldIndex; i++)
        {
            oSheet.getCellByPosition (i, 0).setFormula ("Col" + i);
            oSheet.getCellByPosition (0, i).setFormula ("Row" + i);
            oSheet2.getCellByPosition (i, 0).setFormula ("Col" + i);
            oSheet2.getCellByPosition (0, i).setFormula ("Row" + i);
        }

        for (int i = 1; i < mMaxFieldIndex; i++)
        {
            for (int j = 1; j < mMaxFieldIndex; j++)
            {
                oSheet.getCellByPosition (i, j).setValue (i * (j + 1));
                oSheet2.getCellByPosition (i, j).setValue (i * (j + 2));
            }
        }

        oSheet.getCellByPosition (1, 1).setFormula ("aName");
        oSheet.getCellByPosition (1, 2).setFormula ("otherName");
        oSheet.getCellByPosition (1, 3).setFormula ("una");
        oSheet.getCellByPosition (1, 4).setFormula ("otherName");
        oSheet.getCellByPosition (1, 5).setFormula ("somethingelse");

        oSheet.getCellByPosition (1, 5);

        int x = sCellAdress.Column;
        int y = sCellAdress.Row + 3;

        oSheet.getCellByPosition (x, y);


        // create the test objects
        log.println ("Getting test objects");

        XDataPilotTablesSupplier DPTS = UnoRuntime.queryInterface (
            XDataPilotTablesSupplier.class,
            oSheet);
        XDataPilotTables DPT = DPTS.getDataPilotTables ();
        XDataPilotDescriptor DPDsc = DPT.createDataPilotDescriptor ();
        DPDsc.setSourceRange (sCellRangeAdress);

        XPropertySet fieldPropSet = null;

        Object oDataPilotField = DPDsc.getDataPilotFields ().getByIndex (0);
        fieldPropSet = UnoRuntime.queryInterface (
            XPropertySet.class, oDataPilotField);
        fieldPropSet.setPropertyValue ("Orientation",
            com.sun.star.sheet.DataPilotFieldOrientation.ROW);
        oDataPilotField = DPDsc.getDataPilotFields ().getByIndex (1);
        fieldPropSet = UnoRuntime.queryInterface (
            XPropertySet.class, oDataPilotField);
        fieldPropSet.setPropertyValue ("Function",
            com.sun.star.sheet.GeneralFunction.SUM);
        fieldPropSet.setPropertyValue ("Orientation",
            com.sun.star.sheet.DataPilotFieldOrientation.DATA);
        oDataPilotField = DPDsc.getDataPilotFields ().getByIndex (2);
        fieldPropSet = UnoRuntime.queryInterface (
            XPropertySet.class, oDataPilotField);
        fieldPropSet.setPropertyValue ("Orientation",
            com.sun.star.sheet.DataPilotFieldOrientation.COLUMN);

        log.println ("Insert the DataPilotTable");

        if (DPT.hasByName ("DataPilotTable"))
        {
            DPT.removeByName ("DataPilotTable");
        }

        DPT.insertNewByName ("DataPilotTable", sCellAdress, DPDsc);
        XIndexAccess xIA = UnoRuntime.queryInterface (XIndexAccess.class,DPTS.getDataPilotTables ());
        XIndexAccess IA = null;
        XDataPilotDescriptor xDPT = UnoRuntime.queryInterface (XDataPilotDescriptor.class,xIA.getByIndex (0));
        IA = xDPT.getRowFields ();
        System.out.println ("COUNT: "+IA.getCount ());
        datapilotfield = (XInterface) AnyConverter.toObject (
            new Type (XInterface.class), IA.getByIndex (0));

        try
        {
            XDataPilotFieldGrouping  dpfg = UnoRuntime.queryInterface (XDataPilotFieldGrouping.class, datapilotfield);
            String[] elements = new String[]{"aName","otherName"};
            dpfg.createNameGroup (elements);
            DataPilotFieldGroupInfo dpgi=null;
            xIA = UnoRuntime.queryInterface (XIndexAccess.class,DPTS.getDataPilotTables ());
            XDataPilotDescriptor xDPT2 = UnoRuntime.queryInterface (XDataPilotDescriptor.class,xIA.getByIndex (0));
            IA = xDPT2.getRowFields ();
            for (int i=0;i<IA.getCount ();i++)
            {
                datapilotfield = (XInterface) AnyConverter.toObject (
                    new Type (XInterface.class), IA.getByIndex (i));
                XPropertySet xPropertySet = UnoRuntime.queryInterface (XPropertySet.class, IA.getByIndex (i));
                if (((Boolean)xPropertySet.getPropertyValue ("IsGroupField")).booleanValue ())
                {
                    xPropertySet = UnoRuntime.queryInterface (XPropertySet.class, datapilotfield);
                    XNamed xNamed = UnoRuntime.queryInterface (XNamed.class, IA.getByIndex (i));
                    System.out.println ("name: "+xNamed.getName ());
                    dpgi = (DataPilotFieldGroupInfo) xPropertySet.getPropertyValue ("GroupInfo");
                }
            }
            groups = dpgi.Groups;
            XIndexAccess groupAccess = UnoRuntime.queryInterface(XIndexAccess.class, groups);
            XNameAccess groupNames = UnoRuntime.queryInterface(XNameAccess.class, groupAccess.getByIndex(0));
            oObj = UnoRuntime.queryInterface(XInterface.class, groupNames.getByName("aName"));
        }
        catch (Exception e)
        {
            e.printStackTrace ();
        }

        log.println ("Creating object - " +
            ((oObj == null) ? "FAILED" : "OK"));

        TestEnvironment tEnv = new TestEnvironment (oObj);

        log.println ("Implementationname: " + util.utils.getImplName (oObj));

        // Other parameters required for interface tests
        return tEnv;
    }

}

File: sc/qa/extras/scdatapilotfieldgroupitemobj.cxx
/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include <test/unoapi_test.hxx>
#include <test/container/xnamed.hxx>
#include <test/lang/xserviceinfo.hxx>

#include <com/sun/star/beans/XPropertySet.hpp>
#include <com/sun/star/container/XIndexAccess.hpp>
#include <com/sun/star/container/XNameAccess.hpp>
#include <com/sun/star/sheet/DataPilotFieldGroupInfo.hpp>
#include <com/sun/star/sheet/DataPilotFieldOrientation.hpp>
#include <com/sun/star/sheet/GeneralFunction.hpp>
#include <com/sun/star/sheet/XDataPilotDescriptor.hpp>
#include <com/sun/star/sheet/XDataPilotFieldGrouping.hpp>
#include <com/sun/star/sheet/XDataPilotTables.hpp>
#include <com/sun/star/sheet/XDataPilotTablesSupplier.hpp>
#include <com/sun/star/sheet/XSpreadsheet.hpp>
#include <com/sun/star/sheet/XSpreadsheetDocument.hpp>
#include <com/sun/star/sheet/XSpreadsheets.hpp>
#include <com/sun/star/table/CellAddress.hpp>
#include <com/sun/star/table/CellRangeAddress.hpp>
#include <com/sun/star/uno/XInterface.hpp>

#include <com/sun/star/uno/Reference.hxx>
#include <com/sun/star/uno/Sequence.hxx>

#include <comphelper/types.hxx>

using namespace css;

namespace sc_apitest
{
class ScDataPilotFieldGroupItemObj : public UnoApiTest,
                                     public apitest::XNamed,
                                     public apitest::XServiceInfo
{
public:
    ScDataPilotFieldGroupItemObj();

    virtual uno::Reference<uno::XInterface> init() override;
    virtual void setUp() override;

    CPPUNIT_TEST_SUITE(ScDataPilotFieldGroupItemObj);

    // XNamed
    CPPUNIT_TEST(testGetName);
    CPPUNIT_TEST(testSetName);

    // XServiceInfo
    CPPUNIT_TEST(testGetImplementationName);
    CPPUNIT_TEST(testGetSupportedServiceNames);
    CPPUNIT_TEST(testSupportsService);

    CPPUNIT_TEST_SUITE_END();

private:
    static const int m_nMaxFieldIndex = 6;
};

ScDataPilotFieldGroupItemObj::ScDataPilotFieldGroupItemObj()
    : UnoApiTest(u"/sc/qa/extras/testdocuments"_ustr)
    , XNamed(u"aName"_ustr)
    , XServiceInfo(u"ScDataPilotFieldGroupItemObj"_ustr,
                   u"com.sun.star.sheet.DataPilotFieldGroupItem"_ustr)
{
}

uno::Reference<uno::XInterface> ScDataPilotFieldGroupItemObj::init()
{
    table::CellRangeAddress aCellRangeAddress(0, 1, 0, m_nMaxFieldIndex - 1, m_nMaxFieldIndex - 1);
    table::CellAddress aCellAddress(0, 7, 8);

    uno::Reference<sheet::XSpreadsheetDocument> xDoc(mxComponent, uno::UNO_QUERY_THROW);
    uno::Reference<sheet::XSpreadsheets> xSheets(xDoc->getSheets(), uno::UNO_SET_THROW);
    uno::Reference<container::XIndexAccess> xIA(xSheets, uno::UNO_QUERY_THROW);
    xSheets->insertNewByName(u"Some Sheet"_ustr, 0);

    uno::Reference<sheet::XSpreadsheet> xSheet0(xIA->getByIndex(0), uno::UNO_QUERY_THROW);
    uno::Reference<sheet::XSpreadsheet> xSheet1(xIA->getByIndex(1), uno::UNO_QUERY_THROW);

    for (auto i = 1; i < m_nMaxFieldIndex; ++i)
    {
        xSheet0->getCellByPosition(i, 0)->setFormula("Col" + OUString::number(i));
        xSheet0->getCellByPosition(0, i)->setFormula("Row" + OUString::number(i));
        xSheet1->getCellByPosition(i, 0)->setFormula("Col" + OUString::number(i));
        xSheet1->getCellByPosition(0, i)->setFormula("Row" + OUString::number(i));
    }

    for (auto i = 1; i < m_nMaxFieldIndex; ++i)
    {
        for (auto j = 1; j < m_nMaxFieldIndex; ++j)
        {
            xSheet0->getCellByPosition(i, j)->setValue(i * (j + 1));
            xSheet1->getCellByPosition(i, j)->setValue(i * (j + 2));
        }
    }

    xSheet0->getCellByPosition(1, 1)->setFormula(u"aName"_ustr);
    xSheet0->getCellByPosition(1, 2)->setFormula(u"otherName"_ustr);
    xSheet0->getCellByPosition(1, 3)->setFormula(u"una"_ustr);
    xSheet0->getCellByPosition(1, 4)->setFormula(u"otherName"_ustr);
    xSheet0->getCellByPosition(1, 5)->setFormula(u"somethingelse"_ustr);

    xSheet0->getCellByPosition(1, 5);
    xSheet0->getCellByPosition(aCellAddress.Column, aCellAddress.Row + 3);

    uno::Reference<sheet::XDataPilotTablesSupplier> xDPTS(xSheet0, uno::UNO_QUERY_THROW);
    uno::Reference<sheet::XDataPilotTables> xDPT(xDPTS->getDataPilotTables(), uno::UNO_SET_THROW);
    uno::Reference<sheet::XDataPilotDescriptor> xDPD(xDPT->createDataPilotDescriptor(),
                                                     uno::UNO_SET_THROW);

    xDPD->setSourceRange(aCellRangeAddress);

    uno::Reference<beans::XPropertySet> xPropertySet0(xDPD->getDataPilotFields()->getByIndex(0),
                                                      uno::UNO_QUERY_THROW);
    xPropertySet0->setPropertyValue(u"Orientation"_ustr,
                                    uno::Any(sheet::DataPilotFieldOrientation_ROW));

    uno::Reference<beans::XPropertySet> xPropertySet1(xDPD->getDataPilotFields()->getByIndex(1),
                                                      uno::UNO_QUERY_THROW);
    xPropertySet1->setPropertyValue(u"Function"_ustr, uno::Any(sheet::GeneralFunction_SUM));
    xPropertySet1->setPropertyValue(u"Orientation"_ustr,
                                    uno::Any(sheet::DataPilotFieldOrientation_DATA));

    uno::Reference<beans::XPropertySet> xPropertySet2(xDPD->getDataPilotFields()->getByIndex(2),
                                                      uno::UNO_QUERY_THROW);
    xPropertySet2->setPropertyValue(u"Orientation"_ustr,
                                    uno::Any(sheet::DataPilotFieldOrientation_COLUMN));

    xDPT->insertNewByName(u"DataPilotTable"_ustr, aCellAddress, xDPD);

    uno::Reference<container::XIndexAccess> xIA_DPT0(xDPTS->getDataPilotTables(),
                                                     uno::UNO_QUERY_THROW);
    uno::Reference<sheet::XDataPilotDescriptor> xDPD0(xIA_DPT0->getByIndex(0),
                                                      uno::UNO_QUERY_THROW);
    uno::Reference<container::XIndexAccess> xIA_RF0(xDPD0->getRowFields(), uno::UNO_SET_THROW);

    uno::Reference<sheet::XDataPilotFieldGrouping> xDPFG(xIA_RF0->getByIndex(0),
                                                         uno::UNO_QUERY_THROW);
    xDPFG->createNameGroup({ u"aName"_ustr, u"otherName"_ustr });

    uno::Reference<container::XIndexAccess> xIA_DPT1(xDPTS->getDataPilotTables(),
                                                     uno::UNO_QUERY_THROW);
    uno::Reference<sheet::XDataPilotDescriptor> xDPD1(xIA_DPT1->getByIndex(0),
                                                      uno::UNO_QUERY_THROW);
    uno::Reference<container::XIndexAccess> xIA_RF1(xDPD1->getRowFields(), uno::UNO_SET_THROW);

    sheet::DataPilotFieldGroupInfo aDPFGI;
    for (auto i = 0; i < xIA_RF1->getCount(); ++i)
    {
        uno::Reference<beans::XPropertySet> xPropertySet(xIA_RF1->getByIndex(i),
                                                         uno::UNO_QUERY_THROW);
        if (::comphelper::getBOOL(xPropertySet->getPropertyValue(u"IsGroupField"_ustr)))
        {
            CPPUNIT_ASSERT(xPropertySet->getPropertyValue(u"GroupInfo"_ustr) >>= aDPFGI);
        }
    }

    uno::Reference<container::XIndexAccess> xIA_GI(aDPFGI.Groups, uno::UNO_QUERY_THROW);
    uno::Reference<container::XNameAccess> xNA_GN(xIA_GI->getByIndex(0), uno::UNO_QUERY_THROW);

    uno::Reference<uno::XInterface> xReturn(xNA_GN->getByName(u"aName"_ustr), uno::UNO_QUERY_THROW);
    return xReturn;
}

void ScDataPilotFieldGroupItemObj::setUp()
{
    UnoApiTest::setUp();
    // create a calc document
    mxComponent = loadFromDesktop(u"private:factory/scalc"_ustr);
}

CPPUNIT_TEST_SUITE_REGISTRATION(ScDataPilotFieldGroupItemObj);

} // namespace sc_apitest

CPPUNIT_PLUGIN_IMPLEMENT();

/* vim:set shiftwidth=4 softtabstop=4 expandtab: */
